### PR TITLE
dtype when creating elements

### PR DESCRIPTION
**Pull request by mscroggs**
- Default dtype updated to match DOLFINx (if installed)
- dtype input added to custom element

Will allow us to do https://github.com/FEniCS/dolfinx/pull/3136

### CODE REVIEW

**Comment by jhale** [comment]
What's the status of this?

**Comment by garth-wells** [comment]
I'm not wild about bringing DOLFINx into this. I'd favour defaulting to `float64` or being completely explicit on the type.

**Comment by jorgensd** [comment]
> I'm not wild about bringing DOLFINx into this. I'd favour defaulting to `float64` or being completely explicit on the type.

I agree with Garth

**Comment by mscroggs** [comment]
> What's the status of this?

Needs a bit of work before it's ready. Should have time to do this work in the next couple of weeks

**Comment by mscroggs** [comment]
> > What's the status of this?
> 
> Needs a bit of work before it's ready. Should have time to do this work in the next couple of weeks

It was less work to fix that I thought it might be. Ready for review now

File: python/basix/finite_element.py [hunk 1]
```diff
@@ -582,7 +587,7 @@ def create_element(
     dpc_variant: DPCVariant = DPCVariant.unset,
     discontinuous: bool = False,
     dof_ordering: typing.Optional[list[int]] = None,
-    dtype: npt.DTypeLike = np.float64,
+    dtype: typing.Optional[npt.DTypeLike] = None,
```
Comment (garth-wells) [inline python/basix/finite_element.py:590]:
```suggestion
    dtype: npt.DTypeLike =  np.float64,
```

File: python/basix/finite_element.py [hunk 2]
```diff
@@ -602,19 +607,24 @@ def create_element(
...
-            family.value,
-            celltype.value,
-            degree,
-            lagrange_variant.value,
-            dpc_variant.value,
-            discontinuous,
-            dof_ordering if dof_ordering is not None else [],
-            np.dtype(dtype).char,
-        )
+    if dtype is None:
```
Comment (garth-wells) [inline python/basix/finite_element.py:610]: Remove?

File: python/basix/finite_element.py [hunk 3]
```diff
@@ -657,26 +668,56 @@ def create_custom_element(
...
+
+        dtype = default_scalar_type
+
+    if wcoeffs.dtype != dtype:
+        wcoeffs = np.dtype(dtype).type(wcoeffs)  # type: ignore
+        x = [[np.dtype(dtype).type(j) for j in i] for i in x]  # type: ignore
+        M = [[np.dtype(dtype).type(j) for j in i] for i in M]  # type: ignore
+    if np.issubdtype(dtype, np.float32):
+        return FiniteElement(
+            _create_custom_element_float32(
```
Comment (garth-wells) [inline python/basix/finite_element.py:687]: I wonder if using `std::variant` in the nanobind layer could avoid the duplication with  `_create_custom_element_float32` and  `_create_custom_element_float64`.

File: python/basix/finite_element.py (hunk 1 above)
Comment (mscroggs) [reply python/basix/finite_element.py:590]: Doing it the way we currently do it allows us to do this in DOLFINx to make Basix use the same default scalar type, which is convenient:

https://github.com/FEniCS/dolfinx/blob/5332dc9f4b97054a63264d8fa4711ed5e5ff0b1a/python/dolfinx/__init__.py#L39-L41

File: python/basix/finite_element.py (hunk 2 above)
Comment (garth-wells) [reply python/basix/finite_element.py:610]: @mscroggs

Comment (mscroggs) [reply python/basix/finite_element.py:610]: This is needed to allow DOLFINx to change the default scalar type as in the comment above

Comment (garth-wells) [reply python/basix/finite_element.py:610]: This all seems too implicit to me. I think DOLFINx should explicitly pass the type to Basix functions. Global defaults are not nice.

Comment (mscroggs) [reply python/basix/finite_element.py]: ok, default_scalar_type is gone

**Review by garth-wells — APPROVED** [review]

**Pull request merged**